Add Mapped Document Roots functionality (#22)
mezzio/mezzio-swoole: docs/book/v2/static-resources.md

Once implemented, map the service
`Mezzio\Swoole\StaticResourceHandlerInterface` to a factory that
returns your custom implementation within your `dependencies` configuration.

## Example alternate static resource handler: StaticMappedResourceHandler

> 

Reviewer: ```suggestion

- Since 2.7.0

```

So that users understand why it might not be available to them.

> The default static resource handler, `Mezzio\Swoole\StaticResourceHandler`, requires all files to be in specified document root (default "public") directory when instantiating the handler.  If you are using modules generating templates with associated file assets (JavaScript, CSS, etc.), those files have to be copied to the "public" directory to work.  This can be done via scripting, but is one more step to concern about when testing or deploying a site.  Ideally, a module should be able to contain both its template and any dependencies that template relies upon.  

Reviewer: ```suggestion
The default static resource handler, `Mezzio\Swoole\StaticResourceHandler`, requires all files to be in the specified document root directory (by default, "public") when instantiating the handler.  If you are using modules generating templates with associated file assets (JavaScript, CSS, etc.), those files must be copied to the "public" directory if you wish to allow access to them. This can be done via scripting, but it is one more step to consider when testing or deploying a site. Ideally, a module should be able to contain both its template and any dependencies that template relies upon.
```

(Minor improvements.)

For example, assume you have a module, AwesomeModule, with a handler called "HomeHandler", which renders the 'home' template.  You designate the prefix, `/awesome-home` for rendering the assets.  The structure of your module files looks like this:

> ```

Reviewer: ```suggestion
```text
```

This is required so that it gets rendered correctly as a code block.

AwesomeModule
├── src
|   ├── Handler
|   |   ├── HomeHandler.php
|   |   ├── HomeHandlerFactory.php
|   ├── ConfigProvider.php
├── templates
│   ├── home
|   |   ├── home.html
|   |   ├── style.css
│   ├── layouts
```

In your `home.html` template, you can refer to the `style.css` file, using `/awesome-home` as follows:

```html
<link href="/awesome-home/style.css" rel="stylesheet" type="text/css">
```

> 

Reviewer: The next section dives directly into describing usage of the new implementation... but we haven't been introduced to it yet. Addtionally, the above section details how you _want_ it to work, but not how it _currently_ works under the default implementation.

I'd add some verbiage along the lines of:

> Currently, however, this will cause errors, as the stylesheet will not be available under the `public/` tree of the application. You would need to copy or symlink the files to the appropriate location to make that work.
>
> To solve this problem, we have introduced an additional static resource handler implementation, the `StaticMappedResourceHandler`.

From there, we can go into usage.

### Using StaticMappedResourceHandler

> To use StaticMappedResourceHandler from an application or module:

Reviewer: ```suggestion
To use `Mezzio\Swoole\StaticMappedResourceHandler` from an application or module:
```

> 
> 1. Define what your URI prefix will be (ex. `/awesome-home`)

Reviewer: ```suggestion
1. Define what your URI prefix will be (e.g. `/awesome-home`)
```

> 2. Update your template/s attributes like `href` and `src` to use the prefix (ex. `<script src='/awesome-home/style.css'></script>`)

Reviewer: ```suggestion
2. Update references to linkable resources in your templates to use the desired prefix (e.g. `<script src='/awesome-home/style.css'></script>`)
```

> 3. In your application/module configuration (or ConfigProvider), add the relationship between your prefix (`awesome-home`) and the directory/ies containing the assets

Reviewer: ```suggestion
3. In your application/module configuration (or `ConfigProvider`), add the relationship between your prefix (`awesome-home`) and any directories containing the assets
```

Also, indicate that there is an example of this configuration below.

> 4. In the application's configuration, set the alias of `Mezzio\Swoole\StaticResourceHandlerInterface` to use `Mezzio\Swoole\StaticMappedResourceHandler`

Reviewer: Indicate here that there is an example below demonstrating this.

In step #4, in your application's configuration (`autoload/dependencies.global.php` is a good place), override the default implementation of StaticResourceHandlerInterface:

```php
return [
    'dependencies' => [
        'aliases' => [
            Mezzio\Swoole\StaticResourceHandlerInterface::class => Mezzio\Swoole\StaticMappedResourceHandler::class
            // Fully\Qualified\ClassOrInterfaceName::class => Fully\Qualified\ClassName::class,
        ],
        // etc.
```

For step #3, an alternative to storing a configuration is dynamically associating `/awesome-home` to a directory in code (probably within a factory).  
This approach could be useful if the directory of the assets isn't know until runtime.

```php
use Psr\Container\ContainerInterface;
use Mezzio\Template\TemplateRendererInterface;
use Mezzio\Swoole\StaticResourceHandler\FileLocationRepositoryInterface;

class AwesomeHomeHandlerFactory
{
    public function __invoke(ContainerInterface $container) : DocumentationViewHandler
    {
        // Establish location for the home template assets
        $repo = $container->get(FileLocationRepositoryInterface::class);

>         $repo->addMappedDocumentRoot('awesome-home', 
>             realpath(__DIR__ . '/../../templates/home'));

Reviewer: ```suggestion
        $repo->addMappedDocumentRoot(
            'awesome-home', 
            realpath(__DIR__ . '/../../templates/home')
        );
```

To follow with our own CS guidelines.

        return new AwesomeHomeHandler(
            $container->get(TemplateRendererInterface::class)
        );
    }
}
```

> When the template renders, the client will request `/awesome-home/style.css`, which the StaticResourceHandler will now retrieve from the `templates/home` folder of the module.

Reviewer: ```suggestion
When the template renders, the client will request `/awesome-home/style.css`, which the `StaticMappedResourceHandler` will now retrieve from the `templates/home` folder of the module.
```
